### PR TITLE
Issue 837  bug derivative quadratic functional

### PR DESCRIPTION
**Pull request by aringh**
Fix for issue #837. I have also added a test.

### CODE REVIEW

**Comment by aringh** [comment]
Technically: the convex conjugate of the quadratic functional when the operator is 0/`None` should be a translation of the indicator functional on 0, with the value set to the minus the constant.

**Comment by aringh** [comment]
Bumping this. This is up for review :-)

File: odl/solvers/functional/default_functionals.py [hunk 1]
```diff
@@ -1760,8 +1760,13 @@ def convex_conj(self):
 
         """
         if self.operator is None:
-            # Everywhere infinite in this case
-            raise ValueError('convex conjugate not defined without operator')
+            # The convex conjugate of an affine functional is a translation of
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: Perhaps move to docstring?

File: odl/test/solvers/functional/default_functionals_test.py [hunk 2]
```diff
@@ -419,22 +419,31 @@ def test_quadratic_form(space):
...
-    assert isinstance(func.convex_conj, odl.solvers.QuadraticForm)
+    # The convex conjugate is a translation of the IndicatorZero
+    assert isinstance(func_no_operator.convex_conj.functional,
+                      odl.solvers.IndicatorZero)
+    assert func_no_operator.convex_conj.functional.constant == -constant
+
+    # Test with no offset
+    func_no_offset = odl.solvers.QuadraticForm(operator, constant=constant)
+    expected_result = x.inner(operator(x)) + constant
+    assert almost_equal(func_no_offset(x), expected_result)
```
Comment (adler-j) [inline odl/test/solvers/functional/default_functionals_test.py]: `almost_equal` should be replaced with `pytest.approx` accross the board.

File: odl/test/solvers/functional/default_functionals_test.py [hunk 3]
```diff
@@ -419,22 +419,31 @@ def test_quadratic_form(space):
...
-    assert all_almost_equal(func.gradient(x), expected_gradient)
+    expected_gradient = vector
+    assert all_almost_equal(func_no_operator.gradient(x), expected_gradient)
 
-    # The convex conjugate
-    assert isinstance(func.convex_conj, odl.solvers.QuadraticForm)
+    # The convex conjugate is a translation of the IndicatorZero
+    assert isinstance(func_no_operator.convex_conj.functional,
+                      odl.solvers.IndicatorZero)
+    assert func_no_operator.convex_conj.functional.constant == -constant
```
Comment (adler-j) [inline odl/test/solvers/functional/default_functionals_test.py]: would prefer if this was tested by actually evaluating the functional

File: odl/test/solvers/functional/default_functionals_test.py [hunk 4]
```diff
@@ -419,22 +419,31 @@ def test_quadratic_form(space):
...
-    expected_result = x.inner(operator(x)) + constant
-    assert almost_equal(func_no_offset(x), expected_result)
+    # The gradient
+    expected_gradient = 2 * operator(x) + vector
+    assert all_almost_equal(func.gradient(x), expected_gradient)
+
+    # The convex conjugate
+    assert isinstance(func.convex_conj, odl.solvers.QuadraticForm)
 
+    # Test for linear functional Also test with some values as none
```
Comment (adler-j) [inline odl/test/solvers/functional/default_functionals_test.py]: random capitalization of Also

File: odl/test/solvers/functional/default_functionals_test.py (hunk 2 above)
Comment (aringh) [inline odl/test/solvers/functional/default_functionals_test.py]: For the `all_almost_equal` too?

**Comment by aringh** [comment]
If you think that it looks ok, you can merge :-) (I cannot: `You're not authorized to merge this pull request. `)

**Comment by aringh** [comment]
I think that only main developers @adler-j and @kohr-h can merge pull requests
> This pull request can be automatically merged by project collaborators
>Only those with write access to this repository can merge pull requests.
>
>You're not authorized to merge this pull request.
